Switch traceflow CRD validation to webhook validation.

shi0rik0 opened the pull request:
Currently, the traceflow CRD validation is executed in run-time, which is less user-friendly than the webhook validation. I moved most of the validation to the webhook validation.

Some of the validations in Agent is unchanged because they need the information of the Agent. Also, I found the current validation is insufficient, but I left the improvement as a future work.

luolanzone commented on pkg/controller/traceflow/validate_test.go:
```diff
@@ -0,0 +1,88 @@
+		deniedReason string
+	}{
+		{
+			name: "Validating destination IP",
```
I think you also need a success case. You can have a `expectedResponse` field instead of `deniedReason`, please take the Egress validator and test as a reference.

shi0rik0 commented:
`git pull --rebase upstream main` polluted the comparision. Please compare commit [a6aa6d9](https://github.com/antrea-io/antrea/commit/a6aa6d9d03ca2f8fb90500690ab8c22f368f08c9) with [ec9d9ba](https://github.com/antrea-io/antrea/commit/ec9d9ba520bcc54b23001f55fa991e381ef468b0)

shi0rik0 commented:
https://github.com/antrea-io/antrea/compare/ec9d9ba520bcc54b23001f55fa991e381ef468b0..a6aa6d9d03ca2f8fb90500690ab8c22f368f08c9

tnqn commented on pkg/controller/traceflow/validate.go:
```diff
@@ -0,0 +1,111 @@
+		return newResponse(allowed, deniedReason)
+	case admv1.Delete:
+		// This shouldn't happen with the webhook configuration we include in the Antrea YAML manifests.
+		klog.InfoS("Validating DELETE request for Traceflow. This shouldn't happen with the webhook configuration we include in the Antrea YAML manifests.")
```
The second sentence is not proper to be logged. The first one is also too verbose to be V(0). I think either remove this branch and let it fall back to default or just make it V(2) like other two cases.

shi0rik0 replied:
I've merged this branch to the "default" branch.

tnqn commented on pkg/controller/traceflow/validate.go:
```diff
@@ -0,0 +1,111 @@
+		klog.InfoS("Validating DELETE request for Traceflow. This shouldn't happen with the webhook configuration we include in the Antrea YAML manifests.")
+		return newResponse(true, "")
+	default:
+		klog.InfoS("Unknown request type for Traceflow")
```
Even this catch any request, the log won't be helpful to understand what happened. I think ether use klog.ErrorS and add more information about the request, e.g. the actual Operation it receives. or just remove the log, assuming the path is unreachable.

shi0rik0 replied:
I've replaced it with klog.ErrorS.

tnqn commented on pkg/controller/traceflow/validate.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+func (c *Controller) validateCreate(tf *crdv1alpha1.Traceflow) (allowed bool, deniedReason string) {
+	if tf.Spec.Destination.Service != "" && !features.DefaultFeatureGate.Enabled(features.AntreaProxy) {
```
We don't require AntreaProxy to be enabled for antrea-controller, so this won't reflect whether the feature is really enabled in antrea-agent.
https://github.com/antrea-io/antrea/blob/main/docs/feature-gates.md#list-of-available-features

Since there is still a validation in agent, this can be removed.

shi0rik0 replied:
I've moved this validation back to the Agent.

tnqn commented on pkg/controller/traceflow/validate.go:
```diff
@@ -0,0 +1,111 @@
+	if tf.Spec.Destination.Service != "" && !features.DefaultFeatureGate.Enabled(features.AntreaProxy) {
+		return false, "using Service destination requires AntreaProxy feature enabled"
+	}
+	if tf.Spec.Destination.IP != "" {
+		destIP := net.ParseIP(tf.Spec.Destination.IP)
+		if destIP == nil {
+			return false, fmt.Sprintf("destination IP is not valid: %s", tf.Spec.Destination.IP)
+		}
+	}
```
It has been validated via OpenAPI schema: https://github.com/antrea-io/antrea/blob/03d3aeb3b36bf4cf47aae576b903130e18ad7c16/build/charts/antrea/crds/traceflow.yaml#L81-L85
Having this alone may look strange why only destination IP is checked. I think we can just remove it.

tnqn commented on pkg/controller/traceflow/validate.go:
```diff
@@ -0,0 +1,111 @@
+		}
+	}
+	if !tf.Spec.LiveTraffic {
+		srcPod, err := c.podLister.Pods(tf.Spec.Source.Namespace).Get(tf.Spec.Source.Pod)
```
It should first check if Namespace and Pod is empty, and return message like source Pod must be specified in non-live-traffic Traceflow if yes.

shi0rik0 replied:
I have taken note of it. Perhaps we can improve the validation logic in the next PR. For this PR, can we keep the related logic unchanged for now?

tnqn replied:
ditto

tnqn commented on pkg/controller/traceflow/validate.go:
```diff
@@ -0,0 +1,111 @@
+	return true, ""
+}
+
+func (c *Controller) validateUpdate(oldTf, newTf *crdv1alpha1.Traceflow) (allowed bool, deniedReason string) {
+	return c.validateCreate(newTf)
+}
```
for better readability and simplicity, you can rename `validateCreate` to `validate` or `validateTraceflow`, and call it in `Validate` directly. There is no point to decode `oldObj` since it's never used.

shi0rik0 replied:
My consideration is that the validation logic for updates and creations is indeed different because when validating an update, we may need to consider whether the spec before and after the modification are consistent. Or I think maybe we can prohibit updating traceflow, as it seems to make no sense.

tnqn replied:
If they could be different, it makes less sense to call validateCreate in validateUpdate? 

You can always change the code and decode the old object when implementing more validation. But if you just have that in mind and never implement it, some useless code will be kept there perhaps forever and may confuse other contributors, can we keep only necessary code in this commit?

shi0rik0 replied:
Indeed it makes no sense to call validateCreate in validateUpdate, so I think the original logic has problems, but in this PR, I prefer to keep the logic unchanged.

Of course we can keep the lines of code as small as possible, but I think a little bit of "over-design" won't cause confusion.

tnqn commented on pkg/controller/traceflow/validate_test.go:
```diff
@@ -0,0 +1,175 @@
+	crdv1alpha1 "antrea.io/antrea/pkg/apis/crd/v1alpha1"
+)
+
+func TestControllerValidateTraceflow(t *testing.T) {
```
```suggestion
func TestControllerValidate(t *testing.T) {
```
to echo the test target `controller.Validate`

tnqn commented on pkg/controller/traceflow/validate_test.go:
```diff
@@ -0,0 +1,175 @@
+					Operation: admv1.Create,
+					Object:    toRawExtension(tc.newSpec),
+				}
+			} else {
+				t.Fatal("invalid test case")
+			}
```
The input is under control, seems too verbose to handle impossible cases.

shi0rik0 replied:
When we add test cases, it may happen that the added test case is invalid. Reporting this may help debugging. I don't think three lines of validation would be too verbose.

tnqn replied:
This is test for test, which is indeed verbose IMO. It may be only 3 lines in this test, but the same can apply to all tests. If we do such validation for all test cases, there could be thousands of lines and more. I have never seen a project adds tests for tests in practice.
To keep code clean, I would suggest only to assume error for uncontrollable cases, like when it makes an external call.

shi0rik0 replied:
I've deleted it. But still I have some opinions: sure, we cannot gurantee that code is 100% safe, or we will have tons of meta-test, meta-meta-test, and so on. However, this doesn't mean that trivial tests are meaningless. It looks like an "convention documentation" to me rather than an redundant test. It reminds me that such invalid test cases should never exist.

tnqn commented on pkg/controller/traceflow/validate_test.go:
```diff
@@ -0,0 +1,175 @@
+	return runtime.RawExtension{Raw: raw}
+}
+
+func createPods(controller *traceflowController, pods []v1.Pod) {
+	if pods == nil {
+		return
+	}
+	for _, pod := range pods {
+		ns := pod.ObjectMeta.Namespace
+		name := pod.ObjectMeta.Name
+		controller.kubeClient.CoreV1().Pods(ns).Create(context.TODO(), &pod, metav1.CreateOptions{})
+		_, err := controller.waitForPodInNamespace(ns, name, time.Second)
+		if err != nil {
+			panic(err)
+		}
+	}
+}
```
You can pass `pods` to `newController` directly.
```
controller := newController(tc.pods...)
```

shi0rik0 commented on pkg/controller/traceflow/validate_test.go:
```diff
@@ -0,0 +1,161 @@
+	tf := &crdv1alpha1.Traceflow{Spec: *spec}
+	tf.Name = "tf"
+	raw, err := json.Marshal(tf)
+	if err != nil {
+		panic(err)
+	}
```
ditto, I've deleted it.

tnqn reviewed (commented):
LGTM. However, conflicts need to be resolved.

tnqn commented:
> LGTM. However, conflicts need to be resolved.

Ignore it, I mistook other message as conflicts.

shi0rik0 commented:
I've removed outdated unit tests. Interestingly, the unit test "timeoutHostnetworkTraceflow" in pkg/controller/traceflow/controller_test.go seems to be wrong before this patch. The `timeout` in the test case is 2 secs, so we need to wait for at least 2 secs to get the `Failed` status. I don't know why originally the test case can be passed.

Since I don't have the permission to run e2e tests, could you trigger them for me? @tnqn

shi0rik0 commented:
There's still failed tests in "Go / Unit test (windows-2019)", but this failure seems to have nothing to do with this patch?

tnqn commented:
> I've removed outdated unit tests. Interestingly, the unit test "timeoutHostnetworkTraceflow" in pkg/controller/traceflow/controller_test.go seems to be wrong before this patch. The `timeout` in the test case is 2 secs, so we need to wait for at least 2 secs to get the `Failed` status. I don't know why originally the test case can be passed.

It's correct before the patch. It can fail immediately once it finds out the source Pod is in hostnetwork, instead of waiting 2s. Since you removed the source Pod hostnetwork check in it, it then waits until timeout, which is actually the same as the `timeoutTraceflow`, and it's no longer a possible use case. I think it should be removed.

luolanzone commented on pkg/controller/traceflow/validate.go:
```diff
@@ -0,0 +1,90 @@
+func (c *Controller) validate(tf *crdv1alpha1.Traceflow) (allowed bool, deniedReason string) {
+	if !tf.Spec.LiveTraffic {
+		if tf.Spec.Source.Namespace == "" || tf.Spec.Source.Pod == "" {
+			return false, "source pod must be specified in non-live-traffic Traceflow"
```
```suggestion
			return false, "source Pod must be specified in non-live-traffic Traceflow"
```

shi0rik0 commented:
> > I've removed outdated unit tests. Interestingly, the unit test "timeoutHostnetworkTraceflow" in pkg/controller/traceflow/controller_test.go seems to be wrong before this patch. The `timeout` in the test case is 2 secs, so we need to wait for at least 2 secs to get the `Failed` status. I don't know why originally the test case can be passed.
> 
> It's correct before the patch. It can fail immediately once it finds out the source Pod is in hostnetwork, instead of waiting 2s. Since you removed the source Pod hostnetwork check in it, it then waits until timeout, which is actually the same as the `timeoutTraceflow`, and it's no longer a possible use case. I think it should be removed.

I've removed it.

shi0rik0 commented on pkg/controller/traceflow/validate.go:
```diff
@@ -0,0 +1,90 @@
+	var newObj crdv1alpha1.Traceflow
+	if review.Request.Object.Raw != nil {
+		if err := json.Unmarshal(review.Request.Object.Raw, &newObj); err != nil {
+			klog.ErrorS(err, "Error de-serializing current Traceflow")
```
In this log, since the de-serializing fails, we can't get the traceflow's name. I've added the name in other logs, though.

gran-vmv commented on test/e2e/traceflow_test.go:
```diff
@@ -598,48 +596,6 @@ func testTraceflowIntraNode(t *testing.T, data *TestData) {
-			},
-			expectedPhase:   v1alpha1.Failed,
-			expectedReasons: []string{fmt.Sprintf("Invalid Traceflow request, err: %+v", fmt.Errorf("using hostNetwork Pod as source in non-live-traffic Traceflow is not supported"))},
-		},
```
Could you keep these cases as it can verify the result of invalid request? UT cannot detect the issue on apiserver side.

gran-vmv commented:
/test-e2e

gran-vmv reviewed (dismissed):
LGTM

shi0rik0 commented:
/test-e2e

shi0rik0 commented:
Sorry for the late reply. The conflicts are caused by the Traceflow API version promotion. I've solved the conflicts. There's a failed e2e test, but this seems to be not related to this PR because recent PRs also have this failure (#5380). @luolanzone

tnqn commented on test/e2e/traceflow_test.go:
```diff
@@ -2237,6 +2197,88 @@ func testTraceflowEgress(t *testing.T, data *TestData) {
 }
 
+func testTraceflowValidation(t *testing.T, data *TestData) {
+	createTestAgnhostPodsHostNetwork := func(tb testing.TB, data *TestData, num int, ns string, nodeName string) ([]string, []*PodIPs, func()) {
+		return createTestPods(tb, data, num, ns, nodeName, true, data.createAgnhostPodOnNode)
+	}
+	podNames, podIPs, cleanupFn := createTestAgnhostPodsHostNetwork(t, data, 1, data.testNamespace, nodeName(0))
```
The closure doesn't seem very useful. It could have been one line code but uses 4 lines just to move some parameters to function name.

tnqn reviewed (dismissed):
LGTM, one nit

tnqn reviewed (approved):
LGTM

tnqn commented:
/test-all

shi0rik0 commented:
/test-all

antoninbas commented:
This PR broke e2e tests on IPv6-only clusters, as the new e2e test assumes that Pods always have a valid IPv4 address. I have opened a PR to fix it: https://github.com/antrea-io/antrea/pull/5415

When doing any non-trivial change to e2e Traceflow tests, it's probably a good idea to always run IPv6-only and dual-stack e2e test jobs.
